torch/autograd/_functions/tensor.py

```diff
@@ -565,12 +575,156 @@ def forward(self, input):
+        assume anything on the input. The main downside is
+        that it takes time O(n^2), where n = input.size(self.dim)
+        (i.e. the length of the cumulative product). This is in
+        contrast to the forward pass and the efficient algorithm,
+        which are both O(n).
+
+        The second algorithm is a simple application of the chain
+        rule. If x is an n-dimensional vector, and y = cumprod(x),
+        and F is the final cost, then
+
+        dF / dx_k = sum_j (dF / dy_j) * (dy_j / dx_k)   (1)
+
+        The term dF / dy_j is just grad_output[j] (assuming again
+        everything is one-dimensional).
+
+        The term (dy_j / dx_k) is easilly seen to be
+
+        if j >= k
+            dy_j / dx_k = prod_{1 <= i <= j, i != k} x_i
+        else:
+            dy_j / dx_k = 0
+
+        Note that the indicator (j>=k) can be taken out
+        by replacing the sum in (1) with a sum from
+        j = k to n.
+
+        Thus,
+        df / dx_k = sum_{k <= j <= n} grad_output[j] * (dy_j / dx_k)
+
+        with
+        dy_j / dx_k = prod_{1 <= i <= j, i != k} x_i     (2)
+
+        Note that this last term is just the cumulative product
+        with k omitted. Thus, if x_k (the input) is nonzero, we can
+        just express this as
+
+        dy_j / dx_k = (prod_{1 <= i <= j} x_i) / x_k
+                    = y_j / x_k
+
+        So therefore,
+
+        df / dx_k = sum_{k <= j <= n} grad_output[j] * y_j / x_k
+
+        so
+
+        grad_output = sum_scan_exclusiv(grad_output * output) / input
+
+        If the input is nonzero, we need to calculate the dy_j / dx_k
+        by using the formula (2), called in the code omitted_products.
+
+        The way the code calculates it is simply by noting that
+
+        prod_{1 <= i <= j, i != k} x_i
+            = (prod_{1 <= i <= k} x_i) * (prod_{k + 1 <= i <= j} x_i)
+
+        the first term is calculated as prods_until_k, which since
+        doesn't depend in j is easy to vectorize.
+
+        The second term (indexed by j) is the cumulative product of
+        x_{k+1}, x_{k+2}, ..., x_n, and it's named in the code
+        prods_from_k_pkus_1, and it's calculated as a cumprod.
+
+        In order to vectorize this properly, we need to add to
+        omitted_products the dimensions where k > j, and therefore
+        dy_j / dx_k = 0, which is done right after the assert.
+        '''
+
+        input, = self.saved_tensors
+        dim_size = input.size(self.dim)
+        if dim_size == 1:
+            return grad_output.clone()
+
+        #  Simple case with nonzero elements in the input
+        if (input != 0).all():
+            output = torch.cumprod(input, dim=self.dim)
+            return sum_scan_exclusive(output * grad_output, dim=self.dim) / input
+
+        dim_padding = (slice(None, None),) * self.dim
+
+        ones_size = list(input.size())
+        ones_size[self.dim] = 1
+        ones = input.new().resize_(ones_size).fill_(1)
```

Comment:
I didn't quite get this. Could you elaborate a bit? I need to concatenate the 'ones' tensor later.